[Justin Peh Yu Xiang] iP

pyuxiang opened the pull request.

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

pyuxiang committed:
Add echo command

pyuxiang committed:
Add Tasklist feature

pyuxiang committed:
Replace list items as tasks

pyuxiang committed:
Add ability to mark tasks as done

pyuxiang committed:
Add string method for task

pyuxiang committed:
Add Deadline, Event, ToDos classes

pyuxiang committed:
Add general response methods

pyuxiang committed:
Add task category features

pyuxiang committed:
Add simple test case

pyuxiang committed:
Add custom exceptions and exception handling

pyuxiang committed:
Add delete functionality

pyuxiang committed:
Refactor input processing to use enumeration of commands

pyuxiang committed:
Update test case for enumeration

pyuxiang committed:
Update main to adhere to coding standard

Relevant documentation found here:
https://se-education.org/guides/conventions/java/basic.html

pyuxiang committed:
Update tasks to allow init as done

pyuxiang committed:
Abstract tasklist away from Duke IO parser

pyuxiang committed:
Add method to parse task as entry in file

pyuxiang committed:
Add file loader functionality for task list

pyuxiang committed:
Add datetime parsing functionality

pyuxiang committed:
Merge branch 'branch-Level-7'

pyuxiang committed:
Merge branch 'branch-Level-8'

pyuxiang committed:
Fix bugs from merging

pyuxiang committed:
Refactor classes into nested packages

pyuxiang committed:
Refactor Duke with stricter OOP-style

pyuxiang committed:
Update readability of error messages in Tasks and Commands

pyuxiang committed:
Add tasklist import success message

pyuxiang committed:
Add more datetime parsing options

pyuxiang committed:
Add delete all functionality for tasklist

pyuxiang committed:
Fix bug in delete all requiring two input lines

pyuxiang committed:
Update user messages for more readability

pyuxiang committed:
Add checks for file access/write and enable local read-writes

pyuxiang committed:
Modify Event and Deadline to inherit from DateTask abstraction layer

pyuxiang committed:
Add find functionality to search date range

pyuxiang committed:
Add functionality to view list in sorted order

pyuxiang committed:
Add JUnit tests for Parser and Ui

pyuxiang committed:
Rename data folder for deconflict with common 'data'

pyuxiang committed:
Add manifest

pyuxiang committed:
Add Javadoc comments for public classes and methods

pyuxiang committed:
Update variables as constants

pyuxiang committed:
Change delimiter in file to tabs to avoid conflict with user input

pyuxiang committed:
Update javadoc for rest of the classes

pyuxiang committed:
Update headers to adhere to coding standard

pyuxiang committed:
Add functionality to search using query

pyuxiang committed:
Merge branch 'branch-A-JavaDoc'

pyuxiang committed:
Merge branch 'branch-A-CodingStandard'

pyuxiang committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/commands/DukeCommandFind.java

JoelHo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,73 @@
+import duke.commands.DukeCommand;
+import duke.exceptions.*;
```
Should this be listed explicitly? You can change your IntelliJ preferences to do so!

pyuxiang replied:
Nice tip thanks!

JoelHo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,73 @@
+import duke.tasks.TaskList;
+import duke.ui.Ui;
+
 public class Duke {
```
Should there be a header comment for this? Same with a few other classes too.

pyuxiang replied:
Yes. Barely fulfilling minimal requirements for this.

JoelHo commented on src/main/java/duke/commands/DukeCommand.java:
```diff
@@ -0,0 +1,70 @@
+        if (line.contains(" ")) {
+            int splitIdx = line.indexOf(' ');
+            cmd = line.substring(0, splitIdx); // automatically stripped
+            arg = line.substring(splitIdx+1).strip();
```
Should the + be separated by spaces? A few other instances of this around too.

pyuxiang replied:
Yes

JoelHo commented on src/main/java/duke/commands/DukeCommand.java:
```diff
@@ -0,0 +1,70 @@
+        Task task;
+        switch (cmd) {
+        case "bye":
+            return new DukeCommandBye();
```
Not a coding standard violation, but this is a cyclic dependency, perhaps there's another way to do this? (Not that this is bad but just a comment!)

pyuxiang replied:
Good comment actually. This whole SRP and cyclic dependency is quite problematic for me.
I've recently updated the codebase to use Factory method for a similar issue with `Task <- Todo, Deadline, Event` and replacing a `Task.parseStringAsTask` with:

1. `TaskFileStringParser <- EventFileStringParser`: Offload parsing
2. `GetFileStringParser`: To select appropriate parser (dependency on tasks)

which spawned several more files, a whole lot of message passing, as well as additional javadoc requirements to fulfill. Not sure if this overhead is really worth the effort to decouple the classes.

JoelHo commented on src/main/java/duke/commands/DukeCommand.java:
```diff
@@ -0,0 +1,70 @@
+        }
+    }
+
+    public abstract void execute(TaskList tasks, Ui ui, FileLoader loader) throws DukeException;
```
Should there be a javadoc for this? Then the other commands can inherit the documentation too!

pyuxiang replied:
Yep

JoelHo commented on src/main/java/duke/commands/DukeCommandDelete.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, FileLoader loader)
+            throws DukeExceptionFileNotWritable, DukeExceptionIllegalArgument{
```
Should there be a space before the {?

pyuxiang replied:
Yep

JoelHo commented on src/main/java/duke/commands/DukeCommandFind.java:
```diff
@@ -0,0 +1,149 @@
+    public DukeCommandFind(String input) throws DukeExceptionIllegalArgument {
+        // Process any keywords first, assumed stripped
+        search = "";
+        if (!input.startsWith("/from ")&&(!input.startsWith("/to "))) {
```
Should the && be surrounded by whitespace? Few more instances in this class with operators

pyuxiang replied:
I wasn't aware this is a coding standard thing?

JoelHo commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,251 @@
+                    continue;
+                }
+                date = readout.atTime(0, 0);
+            }
```
Should the else if be on the same line as the close brace? Occurs a few times below too

pyuxiang replied:
Yep, fixed

JoelHo commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,251 @@
+     * @param dt LocalDateTime.
+     * @return Formatted datetime string.
+     */
+    public static String formatDateISO(LocalDateTime dt) {
```
Should the acronym ISO be lowercase (formatDateIso)?

pyuxiang replied:
Yep, great catch, thanks!

JoelHo commented on src/main/java/duke/storage/FileLoader.java:
```diff
@@ -0,0 +1,102 @@
+public class FileLoader {
+
+    protected File f;
+    public boolean isWritable;
+    public boolean isReadable;
```
Should the variables be private? Getters can be written instead if necessary

pyuxiang replied:
Yep

JoelHo commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,88 @@
+ *
+ * A time-based class that tracks a description and a deadline.
+ *
+ * @see duke.tasks.Event
```
I like your extended javadocs!

pyuxiang replied:
Nice

JoelHo commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,88 @@
+     * @param dt LocalDateTime of deadline.
+     * @param isDone Whether task is completed.
+     */
+    public Deadline(String description, LocalDateTime dt, boolean isDone)  {
```
Should there be a singular space before { rather than 2?

JoelHo commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+
+public class TaskList {
+
+    ArrayList<Task> tasks; // composition
```
Should tasks be private? It's not used outside of TaskList either

pyuxiang replied:
(Y)

JoelHo reviewed (commented):
Looks good! Some minor cosmetic issues to be fixed. I really like your javadocs, and the extensive functionality added (especially with the date parsing).

pyuxiang committed:
Merge branch 'add-gradle-support' into master

pyuxiang committed:
Add checkstyle rules

pyuxiang committed:
Fix star import issue with main

pyuxiang committed:
Add JavaFX dependencies to Gradle

pyuxiang committed:
Add JavaFX basic code

pyuxiang committed:
Update JavaFX with simple event listener

pyuxiang committed:
Port Duke into Gui interface

pyuxiang committed:
Update gradle build to enable Gui JAR creation

hengyongming commented on src/main/java/duke/commands/DukeCommand.java:
```diff
@@ -0,0 +1,70 @@
+    public static DukeCommand parse(String line)
+            throws DukeExceptionCommandNotFound, DukeExceptionIllegalArgument {
+        line = line.strip(); // input sanitization
+        String arg = "", cmd = line;
```
Better to separate it into 2 lines

pyuxiang replied:
Yep thanks

hengyongming commented on src/main/java/duke/commands/DukeCommandAdd.java:
```diff
@@ -0,0 +1,31 @@
+    public void execute(TaskList tasks, Ui ui, FileLoader loader) throws DukeExceptionFileNotWritable {
+        tasks.addTask(task);
+        loader.write(tasks);
+        ui.showMessage(
```
May be better to store it in a string variable.

pyuxiang replied:
Store what in a string variable?

hengyongming commented on src/main/java/duke/commands/DukeCommandDone.java:
```diff
@@ -0,0 +1,36 @@
+            throws DukeExceptionFileNotWritable, DukeExceptionIllegalArgument {
+        tasks.setDone(index);
+        loader.write(tasks);
+        ui.showMessage("Nice! I've marked this task as done:", "  "+tasks.getTask(index));
```
There should be a space after  "+"

pyuxiang replied:
Yep

hengyongming commented on src/main/java/duke/commands/DukeCommandFind.java:
```diff
@@ -0,0 +1,149 @@
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, FileLoader loader) throws DukeExceptionIllegalArgument{
+        boolean datedSearch = (from.isPresent() || to.isPresent());
```
searchDate may be a better name

pyuxiang replied:
or even better, `toSearchDate`. Good reco thanks!

hengyongming commented on src/main/java/duke/commands/DukeCommandList.java:
```diff
@@ -0,0 +1,74 @@
+            }
+
+            // Sort in order: Todo < Event (chronological) < Deadline (chronological)
+            view.sort((t1, t2) -> {
```
Store the lambda inside a function?

pyuxiang replied:
Perhaps, but this works pretty well at the moment

hengyongming commented on src/main/java/duke/commands/DukeCommandList.java:
```diff
@@ -0,0 +1,74 @@
+            ui.showMessage("Here are the tasks in your list, in sorted order:", lines);
+
+        } else {
+            ArrayList<String> lines = new ArrayList<>();
```
you could just create an ArrayList at the start to remove repetition

pyuxiang replied:
I'd prefer to avoid initializing this for the `if` case actually, since no processing is performed

pyuxiang replied:
I'd prefer to avoid initializing this for the `if` case actually, since no processing is performed

hengyongming commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,251 @@
+            DateTimeFormatter.ofPattern("H:mm"), // Same/Next day, 24-hour format
+
+            // Date
+            DateTimeFormatter.ofPattern("d/M/yyyy"), // Full date
```
Good effort in catering to all possible input. Perhaps what you can do is to just throw an exception if the input is not in the correct format

pyuxiang replied:
They are throwing exceptions, in `parseDate` below.

hengyongming reviewed (commented):
Well done overall ! Keep it up !

pyuxiang committed:
Add callback for user input by exposing println for Ui

pyuxiang committed:
Fix bug in reading of local tasklist

pyuxiang committed:
Merge branch 'master' into branch-javafx

pyuxiang committed:
Refactor DukeCommand creation to use factory method

pyuxiang committed:
Add comments to DukeCommand.

pyuxiang committed:
Add comments to DukeCommandAdd

pyuxiang committed:
Add comments to DukeCommandBye

pyuxiang committed:
Update DukeCommands to reflect final attributes

pyuxiang committed:
Rename DukeExceptionIllegalCommand and change inheritance

pyuxiang committed:
Add tokenizer for user input

pyuxiang committed:
Fix bugs in tokenization

pyuxiang committed:
Add testcases for user input tokenization

pyuxiang committed:
Clean imports for tokenizer unit test

pyuxiang committed:
Update tokenizer test to test fault inputy

pyuxiang committed:
Rename Parser to more specific DatetimeParser

pyuxiang committed:
Update event to use tokens and include tests

pyuxiang committed:
Refactor inheritance hierarchy for Event

pyuxiang committed:
Remove reference to DatetimeParser from Event

pyuxiang committed:
Refactor tasks to use UserTokens

pyuxiang committed:
Add proper encapsulation for tasks and remove deprecated fileString methods

pyuxiang committed:
Create parser abstractions for tasks

pyuxiang committed:
Refactor entire program to use tokenizer

pyuxiang committed:
Add another assertion for module requirement

pyuxiang committed:
Change method name and imports for code standard.

Many other commit done prior to this commit precisely tackles code quality.

pyuxiang committed:
Highlight lambda usage in code

pyuxiang committed:
Highlight stream usage in code

pyuxiang committed:
Add another code standard.

Code quality improvements already done in the last few commits.

pyuxiang committed:
Merge pull request #1 from pyuxiang/branch-A-Assertions

Add another assertion for module requirement

pyuxiang committed:
Merge remote-tracking branch 'origin/master' into branch-A-CodeQuality

pyuxiang committed:
Merge remote-tracking branch 'origin/master' into branch-A-Lambdas

pyuxiang committed:
Merge remote-tracking branch 'origin/master' into branch-A-Streams

pyuxiang committed:
Merge pull request #2 from pyuxiang/branch-A-CodeQuality

Add another code standard.

pyuxiang committed:
Merge branch 'master' of https://github.com/pyuxiang/ip into branch-A-Streams

pyuxiang committed:
Merge remote-tracking branch 'origin/master' into branch-A-Lambdas

pyuxiang committed:
Merge pull request #3 from pyuxiang/branch-A-Streams

Highlight stream usage in code

pyuxiang committed:
Merge branch 'master' of https://github.com/pyuxiang/ip into branch-A-Lambdas

pyuxiang committed:
Merge pull request #4 from pyuxiang/branch-A-Lambdas

Highlight lambda usage in code

pyuxiang committed:
Add GitHub Actions .yml template

pyuxiang committed:
Empty commit to trigger push for GitHub workflow test

pyuxiang committed:
Fix coding standard and remove IO redirection workflow

pyuxiang committed:
Fix bugs, rectify coding standard violations

pyuxiang committed:
Create Ui abstraction layer and UiGui implementation

pyuxiang committed:
Separate primary and secondary user inputs

pyuxiang committed:
Initial attempt to add simultaneous CLI capabilities

pyuxiang committed:
Replace Ui with FXML specification, build fail

pyuxiang committed:
Add response class

pyuxiang committed:
Add EXIT as possible GUI status

pyuxiang committed:
Redesign GUI to include separate panels for each data stream

Adds additional panels for response by tasklist / error messages, a history log of past commands, and an updated list of the tasklist in the right panel.

pyuxiang committed:
Update panel update and onClick behaviour

pyuxiang committed:
Update border of history textboxes

pyuxiang committed:
Update styling of InputHistoryBox for sleeker look

pyuxiang committed:
Change Duke's responses to Response association class

pyuxiang committed:
Update InputHistoryBox formatting to fade colors after new input

pyuxiang committed:
Change Duke responses from String to Response association class

pyuxiang committed:
Refactor directory structure

Nested main and duke classes into the ./main folder, and respective controllers into ./controllers

pyuxiang committed:
Add external Roboto Mono font

pyuxiang committed:
Show tasklist permanently in sidebar

pyuxiang committed:
Fix delete command bug, indexOutOfBounds

pyuxiang committed:
Fix bug in retrieving exception message for Response

pyuxiang committed:
Add comments and refactoring

pyuxiang committed:
Fix bug in Event done test case

pyuxiang committed:
Update the rest of comments to adhere to coding standard

pyuxiang committed:
Fix import statement in Duke to fulfill check requirements

pyuxiang committed:
Add padding to input history dialog boxes

pyuxiang committed:
Add title and logo

pyuxiang committed:
Update user guide

pyuxiang committed:
Remove outdated test files

pyuxiang committed:
Update user guide to mention /done syntax

pyuxiang committed:
Update readme

pyuxiang committed:
Update readme with mini sample animation

pyuxiang committed:
Add more easter eggs, copy readme to project directory

pyuxiang committed:
Fix task index off-by-one bug in find command

pyuxiang committed:
Align task numbers of the list

Applied to list, find, and Duke tasklist parsing commands

pyuxiang committed:
Adjust wording on readme

pyuxiang committed:
Add newline after logo

pyuxiang committed:
Fix code style to adhere to coding standard

Operator wrapping and whitespace

pyuxiang committed:
Fix resizing issues by disabling resize hint

Noticed problem when textbox was off-aligned, and persisted across startups. Issue due to resizing that shifted bottom anchored text box. Disabling since none of the components are made responsive.

pyuxiang committed:
Replace Ui image with updated one to reflect bugfixes

pyuxiang committed:
Update readme to account for corner cases

Includes erroneous date parsing, formatting, keyboard shortcuts

pyuxiang committed:
Fix grammar error when adding new task(s)

pyuxiang committed:
Add feedback when `find` is not supplied args

pyuxiang committed:
Add fallbacks for monospace font for other OS

pyuxiang committed:
Fix bug to disallow option parsed as a command

Original code uses same parseOption routine to parse commands as well, while allowed commands like `/delete` to be validated as well. Fix is simply to create a new separate parsing routine.

pyuxiang committed:
Merge branch 'master' of https://github.com/pyuxiang/ip

pyuxiang committed:
Update README to reflect project origin disclaimer
